File: src/pl/sda/helloworld/oop/OOPMain.java
package pl.sda.helloworld.oop;

import java.util.Random;

public class OOPMain {

    public static void main(String[] args) {
        new Car(new Engine()).drive();
        new Car(new RacingEngine()).drive();
        new Car(new TruckEngine(false)).drive();


        System.out.println("--------");

        Engine engine = getEngine();
        if(engine instanceof TruckEngine) {
            System.out.println(((TruckEngine) engine).doTruckyThing());
        }
        System.out.println(engine.engineSound());
        System.out.println(engine.toString());
    }

    public static Engine getEngine() {
        if(new Random().nextInt() % 3 == 0) {
            return new RacingEngine();
        } else if (new Random().nextInt() % 3 == 1){
            return new TruckEngine(false);
        } else {
            return new Engine();
        }
    }
}

File: src/pl/sda/helloworld/oop/OOPMain.kt
package pl.sda.helloworld.oop

import com.sun.xml.internal.fastinfoset.util.StringArray

class Main {

        companion object {
            fun main(args: StringArray) {
                val engine = getEngine()
                if(engine is TruckEngine) {
                    println(engine.doTruckyThing())
                }
            }

            fun getEngine(type: String = "") : Engine = TruckEngine(false)

        }
}
